File: dotnet/Mammoth/Couscous/mammoth/internal/util/Lists.cs
using Mammoth.Couscous.java.lang;
using Mammoth.Couscous.java.util;
using Mammoth.Couscous.java.util.function;
using System.Linq;

namespace Mammoth.Couscous.org.zwobble.mammoth.@internal.util {
    internal static class Lists {
        internal static List<T> list<T>(params T[] values) {
            return ToJava.ListToList(values);
        }
        
        internal static List<T> toList<T>(Iterable<T> iterable) {
            return ToJava.ListToList(FromJava.IterableToEnumerable(iterable).ToList());
        }
        
        internal static List<U> eagerMap<T, U>(Iterable<T> iterable, Function<T, U> function) {
            var result = FromJava.IterableToEnumerable(iterable)
                .Select(function.apply)
                .ToList();
            return ToJava.ListToList(result);
        }
        
        internal static List<U> eagerFlatMap<T, U>(Iterable<T> iterable, Function<T, Iterable<U>> function) {
            var result = FromJava.IterableToEnumerable(iterable)
                .SelectMany(element => FromJava.IterableToEnumerable(function.apply(element)))
                .ToList();
            return ToJava.ListToList(result);
        }
        
        internal static List<T> cons<T>(T head, Iterable<T> tail) {
            var result = new T[] { head }
				.Concat(FromJava.IterableToEnumerable(tail))
                .ToList();
			return ToJava.ListToList(result);
        }
        
        internal static List<T> eagerConcat<T>(Iterable<T> first, Iterable<T> second) {
            var result = FromJava.IterableToEnumerable(first)
				.Concat(FromJava.IterableToEnumerable(second))
                .ToList();
			return ToJava.ListToList(result);
        }
        
        internal static List<T> orderedBy<T, R>(Iterable<T> iterable, Function<T, R> getKey) {
            var result = FromJava.IterableToEnumerable(iterable)
                .OrderBy(getKey.apply)
                .ToList();
            return ToJava.ListToList(result);
        }
        
        internal static Optional<T> tryGetLast<T>(List<T> list) {
            if (list.isEmpty()) {
                return new None<T>();
            } else {
                return new Some<T>(list.get(list.size() - 1));
            }
        }
        
        internal static Iterable<T> reversed<T>(List<T> list) {
            return new ReversedList<T>(list);
        }
        
        private class ReversedList<T> : Iterable<T> {
            private readonly List<T> _list;
            
            internal ReversedList(List<T> list) {
                _list = list;
            }
            
            public Iterator<T> iterator() {
                return new ReverseIterator<T>(_list);
            }
        }
        
        private class ReverseIterator<T> : Iterator<T> {
            private readonly List<T> _list;
            private int _nextIndex;
            
            internal ReverseIterator(List<T> list) {
                _list = list;
                _nextIndex = _list.size() - 1;
            }
            
            public bool hasNext() {
                return _nextIndex >= 0;
            }
            
            public T next() {
                if (hasNext()) {
                    return _list.get(_nextIndex--);
                } else {
                    throw new NoSuchElementException();
                }
            }
        }
    }
}

File: src/main/java/org/zwobble/mammoth/internal/util/Lists.java
package org.zwobble.mammoth.internal.util;

import java.util.*;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static org.zwobble.mammoth.internal.util.Iterables.stream;

public class Lists {
    public static <T> List<T> list() {
        return Collections.emptyList();
    }

    public static <T> List<T> list(T value1) {
        return Collections.singletonList(value1);
    }

    public static <T> List<T> list(T value1, T value2) {
        return asList(value1, value2);
    }

    public static <T> List<T> list(T value1, T value2, T value3) {
        return asList(value1, value2, value3);
    }

    @SafeVarargs
    public static <T> List<T> list(T... values) {
        return asList(values);
    }

    public static <T> List<T> cons(T head, Iterable<T> tail) {
        return eagerConcat(list(head), tail);
    }

    public static <T> List<T> eagerConcat(Iterable<T> first, Iterable<T> second) {
        return Stream.concat(stream(first), stream(second)).collect(Collectors.toList());
    }

    public static <T> List<T> eagerFilter(Iterable<T> iterable, Predicate<T> predicate) {
        return stream(iterable).filter(predicate).collect(Collectors.toList());
    }

    public static <T, R> List<R> eagerMap(Iterable<T> iterable, Function<T, R> function) {
        return stream(iterable).map(function).collect(Collectors.toList());
    }

    public static <T, R> List<R> eagerMapWithIndex(Iterable<T> iterable, BiFunction<Integer, T, R> function) {
        return toList(Iterables.lazyMapWithIndex(iterable, function));
    }

    public static <T, R> List<R> eagerFlatMap(Iterable<T> iterable, Function<T, Iterable<R>> function) {
        return stream(iterable)
            .flatMap(element -> stream(function.apply(element)))
            .collect(Collectors.toList());
    }

    public static <T, R extends Comparable<R>> List<T> orderedBy(Iterable<T> iterable, Function<T, R> getKey) {
        return stream(iterable)
            .sorted(orderBy(getKey))
            .collect(Collectors.toList());
    }

    private static <T, R extends Comparable<R>> Comparator<T> orderBy(Function<T, R> getKey) {
        return (first, second) -> getKey.apply(first).compareTo(getKey.apply(second));
    }

    public static <T> List<T> toList(Iterable<T> iterable) {
        return stream(iterable).collect(Collectors.toList());
    }

    public static <T> Optional<T> tryGetLast(List<T> list) {
        if (list.isEmpty()) {
            return Optional.empty();
        } else {
            return Optional.of(list.get(list.size() - 1));
        }
    }

    public static <T> List<T> skip(List<T> list, int count) {
        return list.subList(Math.min(list.size(), count), list.size());
    }

    public static <T> Iterable<T> reversed(List<T> list) {
        return new Iterable<T>() {
            @Override
            public Iterator<T> iterator() {
                ListIterator<T> iterator = list.listIterator(list.size());
                return new Iterator<T>() {
                    @Override
                    public boolean hasNext() {
                        return iterator.hasPrevious();
                    }

                    @Override
                    public T next() {
                        return iterator.previous();
                    }
                };
            }
        };
    }
}
